_specifications/specification-3-17.md

```diff
@@ -8217,6 +8217,182 @@ export interface Moniker {
 
 Server implementations of this method should ensure that the moniker calculation matches to those used in the corresponding LSIF implementation to ensure symbols can be associated correctly across IDE sessions and LSIF indexes.
 
+#### <a href="#textDocument_inlayHints" name="textDocument_inlayHints" class="anchor">Inlay hints (:leftwards_arrow_with_hook:)</a>
+
+> *Since version 3.17.0*
+
+Inlay hints are short textual annotations that are attached to ranges in the source code.
+
+**TODO**: Is "inlay hints" too presentational a name? The protocol allows other UI (e.g. pop-up when cursor is in the target range).
+
+These typically spell out some inferred information, such as the parameter name when passing a value to a function.
+
+```typescript
+/**
+ * Well-known kinds of information conveyed by InlayHints.
+ * This is not an exhaustive list, servers may use other categories as needed.
+ */
+export enum InlayHintCategory {
+	/**
+	 * The range is an expression passed as an argument to a function.
+	 * The label is the name of the parameter.
+	 */
+	'parameter',
+	/**
+	 * The range is an entity whose type is unknown.
+	 * The label is its inferred type.
+	 */
+	'type'
+}
+```
+
+Hints may be grouped into server-defined _classes_.
+Hints in the same class are displayed in similar ways, and the class avoids duplicating this information in each hint.
+
+**TODO**: classes might be useful "subcategories" for users to enable/disable hints. If so, they'd need a description.
+
+**TODO**: are classes actually useful? should we just inline this info into the hints?
+
+```typescript
+export interface InlayHintClass {
+	/**
+	 * The kind of information this hint conveys.
+	 * May be an InlayHintCategory or any other value. 
+	 */
+	category?: string;
```

Comment:
I wouldn't say it's "always" used that way. `Moniker.unique` and `Moniker.kind` use the `enum`s. There are certainly some unexplained discrepancies with how and when enums are used. And there are also several that are defined as `namespace` + `const` + `type`, such as `FailureHandlingKind`.

Reply:
I think that's a different pattern, that exactly illustrates where LSP does/doesn't use the enum types today. Moniker.unique, Moniker.kind, and FailureHandlingKind are (AFAICT) exhaustive, non-enumerated values are not allowed.

Reply:
I'm not clear what you mean. What is the pattern that indicates exhaustiveness? `UniquenessLevel` and `MonikerKind` are `enum`s and `FailureHandlingKind` uses `namespace` + `const` + `type`. The naming scheme of *Kind is similar between two of them, but I don't think that's enough to go on.

Either way, we should not have to rely on assumptions to determine these differences. It's a specification, so if those differences aren't spec'd then it can't be relied upon.

Reply:
The spec says they're exhaustive. It says Moniker.kind has type MonikerKind, an enum type, which in TS's type system means that only the enum values are allowed. Similar with WorkspaceEditClientCapabilities.failureHandling - it's a union type, but with the same effect.

This part of the discussion doesn't seem very relevant to this PR.

I'm fine with `string` + comment or `InlayHintCategory | string`, and this seems like a style question best left to the spec editors.